api/bag_dataset.h

```diff
@@ -0,0 +1,192 @@
+    CompoundLayer& createMetadataProfileCompoundLayer(GeorefMetadataProfile profile,
+                                                      const std::string& name,
+                                                      uint64_t chunkSize, int compressionLevel,
+                                                      DataType keyType = DT_UINT16) &;
+    SurfaceCorrections& createSurfaceCorrections(
+        BAG_SURFACE_CORRECTION_TOPOGRAPHY type, uint8_t numCorrectors,
+        uint64_t chunkSize, int compressionLevel) &;
+    void createVR(uint64_t chunkSize, int compressionLevel, bool makeNode);
+
+    const Metadata& getMetadata() const & noexcept;
+
+    TrackingList& getTrackingList() & noexcept;
+    const TrackingList& getTrackingList() const & noexcept;
+
+    CompoundLayer* getCompoundLayer(const std::string& name) & noexcept;
+    const CompoundLayer* getCompoundLayer(const std::string& name) const & noexcept;
+    std::vector<CompoundLayer*> getCompoundLayers() & noexcept;
+    SurfaceCorrections* getSurfaceCorrections() & noexcept;
+    const SurfaceCorrections* getSurfaceCorrections() const & noexcept;
+    SimpleLayer* getSimpleLayer(LayerType type) & noexcept;
+    const SimpleLayer* getSimpleLayer(LayerType type) const & noexcept;
+
+    VRMetadata* getVRMetadata() & noexcept;
+    const VRMetadata* getVRMetadata() const & noexcept;
+    VRNode* getVRNode() & noexcept;
+    const VRNode* getVRNode() const & noexcept;
+    VRRefinements* getVRRefinements() & noexcept;
+    const VRRefinements* getVRRefinements() const & noexcept;
+    VRTrackingList* getVRTrackingList() & noexcept;
+    const VRTrackingList* getVRTrackingList() const & noexcept;
+
+    Descriptor& getDescriptor() & noexcept;
+    const Descriptor& getDescriptor() const & noexcept;
+
+    std::tuple<double, double> gridToGeo(uint32_t row, uint32_t column) const noexcept;
+    std::tuple<uint32_t, uint32_t> geoToGrid(double x, double y) const noexcept;
+
+private:
+    Dataset() = default;
+    uint32_t getNextId() const noexcept;
+
+    void readDataset(const std::string& fileName, OpenMode openMode);
+    void createDataset(const std::string& fileName, Metadata&& metadata,
+        uint64_t chunkSize, int compressionLevel);
+
+    std::tuple<bool, float, float> getMinMax(LayerType type,
+        const std::string& path = {}) const;
+
+    ::H5::H5File& getH5file() const & noexcept;
+
+    Layer& addLayer(std::unique_ptr<Layer> layer) &;
+
+    //! Custom deleter to not require knowledge of ::H5::H5File destructor here.
+    struct BAG_API DeleteH5File final {
+        void operator()(::H5::H5File* ptr) noexcept;
+    };
+
+    //! The HDF5 file that the BAG is stored in.
+    std::unique_ptr<::H5::H5File, DeleteH5File> m_pH5file;
+    //! The mandatory and optional layers found in the BAG, including ones
+    //! created after opening.
+    std::vector<std::unique_ptr<Layer>> m_layers;
```

Comment:
Many of the various getLayer and similar methods return raw pointers. Should they return smart pointers and if so, would this vector need to contain shared_ptrs instead of unique_ptrs?

Reply:
This has been resolved. Closing.